File: src/lib/export/fit.ts
/**
 * Minimal FIT Activity file writer (FIT Protocol) for F-ACC-003.
 * No third-party SDK — enough for GPX-alternative download in browsers.
 */

import type { Ride } from "@/types";

const CRC_TABLE = (() => {
  const t = new Uint16Array(16);
  for (let i = 0; i < 16; i++) {
    let crc = i;
    for (let j = 0; j < 4; j++) {
      crc = crc & 1 ? 0xcc01 ^ (crc >> 1) : crc >> 1;
    }
    t[i] = crc;
  }
  return t;
})();

function crc16(data: Uint8Array, start = 0, end = data.length): number {
  let crc = 0;
  for (let i = start; i < end; i++) {
    const byte = data[i];
    let tmp = CRC_TABLE[crc & 0xf];
    crc = (crc >> 4) & 0x0fff;
    crc = crc ^ tmp ^ CRC_TABLE[byte & 0xf];
    tmp = CRC_TABLE[crc & 0xf];
    crc = (crc >> 4) & 0x0fff;
    crc = crc ^ tmp ^ CRC_TABLE[(byte >> 4) & 0xf];
  }
  return crc & 0xffff;
}

class FitBuf {
  parts: number[] = [];
  u8(n: number) {
    this.parts.push(n & 0xff);
  }
  u16(n: number) {
    this.parts.push(n & 0xff, (n >> 8) & 0xff);
  }
  u32(n: number) {
    this.parts.push(
      n & 0xff,
      (n >> 8) & 0xff,
      (n >> 16) & 0xff,
      (n >> 24) & 0xff
    );
  }
  i32(n: number) {
    this.u32(n >>> 0);
  }
  bytes(): Uint8Array {
    return new Uint8Array(this.parts);
  }
}

function writeDefinition(
  buf: FitBuf,
  localNum: number,
  globalMsg: number,
  fields: { num: number; size: number; base: number }[]
) {
  buf.u8(0x40 | (localNum & 0x0f));
  buf.u8(0); // reserved
  buf.u8(0); // architecture little
  buf.u16(globalMsg);
  buf.u8(fields.length);
  for (const f of fields) {
    buf.u8(f.num);
    buf.u8(f.size);
    buf.u8(f.base);
  }
}

/** Semi-circles: degrees * (2^31 / 180) */
function toSemi(deg: number): number {
  return Math.round((deg * 0x80000000) / 180);
}

/**
 * Build a minimal FIT activity with record messages from ride.track.
 * Empty track → session/lap only (no synthetic GPS points).
 */
export function rideToFit(ride: Ride): Uint8Array {
  const startMs = new Date(ride.startTime).getTime();
  // FIT epoch = 1989-12-31
  const fitEpoch = Date.UTC(1989, 11, 31, 0, 0, 0);
  const startFit = Math.floor((startMs - fitEpoch) / 1000);

  const pts = ride.track && ride.track.length > 0 ? ride.track : [];

  const records = new FitBuf();
  // file_id definition local 0, global 0
  writeDefinition(records, 0, 0, [
    { num: 0, size: 1, base: 0 }, // type
    { num: 1, size: 2, base: 132 }, // manufacturer
    { num: 2, size: 2, base: 132 }, // product
    { num: 4, size: 4, base: 134 }, // time_created
  ]);
  records.u8(0); // data local 0
  records.u8(4); // activity
  records.u16(255); // development
  records.u16(0);
  records.u32(startFit);

  // record def local 1, global 20
  writeDefinition(records, 1, 20, [
    { num: 253, size: 4, base: 134 }, // timestamp
    { num: 0, size: 4, base: 133 }, // position_lat
    { num: 1, size: 4, base: 133 }, // position_long
    { num: 2, size: 2, base: 132 }, // altitude (scaled)
    { num: 3, size: 1, base: 2 }, // heart_rate
    { num: 4, size: 1, base: 2 }, // cadence
    { num: 7, size: 2, base: 132 }, // power
  ]);

  let written = 0;
  for (let i = 0; i < pts.length; i++) {
    const p = pts[i];
    const lat = typeof p.lat === "number" ? p.lat : null;
    const lng = typeof p.lng === "number" ? p.lng : null;
    // Skip missing / Null Island / out-of-range — never write 0,0.
    if (lat == null || lng == null) continue;
    if (Math.abs(lat) < 1e-6 && Math.abs(lng) < 1e-6) continue;
    if (Math.abs(lat) > 90 || Math.abs(lng) > 180) continue;
    const t =
      typeof p.time === "number"
        ? startFit + Math.floor(p.time)
        : startFit + written * 30;
    records.u8(1);
    records.u32(t);
    records.i32(toSemi(lat));
    records.i32(toSemi(lng));
    const alt = p.elev != null ? Math.round((p.elev + 500) * 5) : 0xffff;
    records.u16(alt);
    const hr = typeof p.hr === "number" && p.hr >= 1 && p.hr <= 239 ? Math.round(p.hr) : 0xff;
    const cad = typeof p.cad === "number" && p.cad >= 1 && p.cad <= 254 ? Math.round(p.cad) : 0xff;
    const power =
      typeof p.power === "number" && p.power >= 1 && p.power <= 2500
        ? Math.round(p.power)
        : 0xffff;
    records.u8(hr);
    records.u8(cad);
    records.u16(power);
    written++;
  }

  // session def local 2, global 18
  writeDefinition(records, 2, 18, [
    { num: 253, size: 4, base: 134 },
    { num: 2, size: 1, base: 0 }, // sport
    { num: 7, size: 4, base: 134 }, // total_elapsed_time (ms/1000 as uint32 * 1000)
    { num: 9, size: 4, base: 134 }, // total_distance (cm)
    { num: 22, size: 2, base: 132 }, // total_ascent
  ]);
  records.u8(2);
  records.u32(startFit + Math.max(1, ride.durationSec));
  records.u8(2); // cycling
  records.u32(Math.round(ride.durationSec * 1000));
  records.u32(Math.round(ride.distanceM * 100));
  records.u16(Math.round(ride.elevationGainM));

  const data = records.bytes();
  const header = new FitBuf();
  header.u8(14); // header size
  header.u8(0x20); // protocol
  header.u16(0x0827); // profile
  header.u32(data.length);
  header.parts.push(...[0x2e, 0x46, 0x49, 0x54]); // .FIT
  const headerBytes = header.bytes();
  const headerCrc = crc16(headerBytes, 0, 12);
  const fullHeader = new Uint8Array(14);
  fullHeader.set(headerBytes.slice(0, 12), 0);
  fullHeader[12] = headerCrc & 0xff;
  fullHeader[13] = (headerCrc >> 8) & 0xff;

  const out = new Uint8Array(14 + data.length + 2);
  out.set(fullHeader, 0);
  out.set(data, 14);
  const fileCrc = crc16(out, 0, 14 + data.length);
  out[14 + data.length] = fileCrc & 0xff;
  out[14 + data.length + 1] = (fileCrc >> 8) & 0xff;
  return out;
}

export function downloadBytes(filename: string, data: Uint8Array, mime: string) {
  if (typeof window === "undefined") return;
  const blob = new Blob([data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength) as ArrayBuffer], {
    type: mime,
  });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

File: mobile/lib/data/export/fit.dart
import 'dart:typed_data';

import '../../domain/ride.dart';

/// Minimal FIT Activity writer — Port of `src/lib/export/fit.ts` (F-ACC-003).
Uint8List rideToFit(RideRecord ride) {
  final startMs = ride.startedAt.toUtc().millisecondsSinceEpoch;
  final fitEpoch = DateTime.utc(1989, 12, 31).millisecondsSinceEpoch;
  final startFit = ((startMs - fitEpoch) / 1000).floor();

  // Empty track → session only (no fake GPS path).
  final pts = ride.track;

  final records = _FitBuf();
  _writeDefinition(records, 0, 0, const [
    (0, 1, 0),
    (1, 2, 132),
    (2, 2, 132),
    (4, 4, 134),
  ]);
  records.u8(0);
  records.u8(4);
  records.u16(255);
  records.u16(0);
  records.u32(startFit);

  _writeDefinition(records, 1, 20, const [
    (253, 4, 134),
    (0, 4, 133),
    (1, 4, 133),
    (2, 2, 132),
    (3, 1, 2),
    (4, 1, 2),
    (7, 2, 132),
  ]);

  var written = 0;
  for (var i = 0; i < pts.length; i++) {
    final p = pts[i];
    final lat = (p['lat'] as num?)?.toDouble();
    final lng = (p['lng'] as num?)?.toDouble() ??
        (p['lon'] as num?)?.toDouble();
    // Skip missing / Null Island / out-of-range — never write 0,0.
    if (lat == null || lng == null) continue;
    if (lat.abs() < 1e-6 && lng.abs() < 1e-6) continue;
    if (lat.abs() > 90 || lng.abs() > 180) continue;
    final elev = (p['elev'] as num?)?.toDouble() ??
        (p['elevation'] as num?)?.toDouble();
    final timeRaw = p['time'] ?? p['timeMs'];
    int t;
    if (timeRaw is num) {
      if (timeRaw > 1e12) {
        t = ((timeRaw.toInt() - fitEpoch) / 1000).floor();
      } else if (timeRaw > 1e10) {
        t = ((timeRaw.toInt() - fitEpoch) / 1000).floor();
      } else {
        t = startFit + timeRaw.round();
      }
    } else {
      t = startFit + written * 30;
    }
    records.u8(1);
    records.u32(t);
    records.i32(_toSemi(lat));
    records.i32(_toSemi(lng));
    final alt = elev != null ? ((elev + 500) * 5).round() : 0xffff;
    records.u16(alt.clamp(0, 0xffff));
    records.u8(liveHrFromTrackPoint(p) ?? 0xff);
    records.u8(liveCadFromTrackPoint(p) ?? 0xff);
    records.u16(livePowerFromTrackPoint(p) ?? 0xffff);
    written++;
  }

  _writeDefinition(records, 2, 18, const [
    (253, 4, 134),
    (2, 1, 0),
    (7, 4, 134),
    (9, 4, 134),
    (22, 2, 132),
  ]);
  records.u8(2);
  records.u32(startFit + (ride.movingTimeSec < 1 ? 1 : ride.movingTimeSec));
  records.u8(2);
  records.u32((ride.movingTimeSec * 1000).round());
  records.u32((ride.distanceKm * 1000 * 100).round());
  records.u16(ride.elevationM.round().clamp(0, 0xffff));

  final data = records.bytes();
  final header = _FitBuf();
  header.u8(14);
  header.u8(0x20);
  header.u16(0x0827);
  header.u32(data.length);
  header.parts.addAll([0x2e, 0x46, 0x49, 0x54]);
  final headerBytes = header.bytes();
  final headerCrc = _crc16(headerBytes, 0, 12);
  final fullHeader = Uint8List(14);
  fullHeader.setRange(0, 12, headerBytes);
  fullHeader[12] = headerCrc & 0xff;
  fullHeader[13] = (headerCrc >> 8) & 0xff;

  final out = Uint8List(14 + data.length + 2);
  out.setRange(0, 14, fullHeader);
  out.setRange(14, 14 + data.length, data);
  final fileCrc = _crc16(out, 0, 14 + data.length);
  out[14 + data.length] = fileCrc & 0xff;
  out[14 + data.length + 1] = (fileCrc >> 8) & 0xff;
  return out;
}

int _toSemi(double deg) => ((deg * 0x80000000) / 180).round();

void _writeDefinition(
  _FitBuf buf,
  int localNum,
  int globalMsg,
  List<(int, int, int)> fields,
) {
  buf.u8(0x40 | (localNum & 0x0f));
  buf.u8(0);
  buf.u8(0);
  buf.u16(globalMsg);
  buf.u8(fields.length);
  for (final f in fields) {
    buf.u8(f.$1);
    buf.u8(f.$2);
    buf.u8(f.$3);
  }
}

final _crcTable = () {
  final t = Uint16List(16);
  for (var i = 0; i < 16; i++) {
    var crc = i;
    for (var j = 0; j < 4; j++) {
      crc = (crc & 1) != 0 ? 0xcc01 ^ (crc >> 1) : crc >> 1;
    }
    t[i] = crc;
  }
  return t;
}();

int _crc16(Uint8List data, int start, int end) {
  var crc = 0;
  for (var i = start; i < end; i++) {
    final byte = data[i];
    var tmp = _crcTable[crc & 0xf];
    crc = (crc >> 4) & 0x0fff;
    crc = crc ^ tmp ^ _crcTable[byte & 0xf];
    tmp = _crcTable[crc & 0xf];
    crc = (crc >> 4) & 0x0fff;
    crc = crc ^ tmp ^ _crcTable[(byte >> 4) & 0xf];
  }
  return crc & 0xffff;
}

class _FitBuf {
  final parts = <int>[];
  void u8(int n) => parts.add(n & 0xff);
  void u16(int n) {
    parts.add(n & 0xff);
    parts.add((n >> 8) & 0xff);
  }

  void u32(int n) {
    parts
      ..add(n & 0xff)
      ..add((n >> 8) & 0xff)
      ..add((n >> 16) & 0xff)
      ..add((n >> 24) & 0xff);
  }

  void i32(int n) => u32(n);

  Uint8List bytes() => Uint8List.fromList(parts);
}
